lib: implement API versioning

lundibundi opened the pull request:
* allow to specify API version during handshake
* make createAppsIndex handle multiple API versions
* add API version parameter to Application
* add API version parameter to SimpleConnectPolicy
* add tests for interface versioning

Fixes: #31.

nechaido commented on lib/applications.js:
```diff
@@ -56,10 +58,39 @@ apps.Application = Application;
+      index.set(application.name, appVersions);
+    }
+    if (!application.version) {
+      // no version means default version
```
I'd change `default` to `latest`, but it is not necessary.

lundibundi replied:
Sure.

nechaido commented:
It should be mentioned that this PR Fixes: #31.

belochub commented on lib/applications.js:
```diff
@@ -56,10 +58,39 @@ apps.Application = Application;
+    }
+    if (!application.version) {
+      // no version means default version
+      if (appVersions.get('latest')) {
```
You should definitely use `has` method here because you don't use the value.

lundibundi replied:
Oh, thanks.

belochub commented on lib/applications.js:
```diff
@@ -56,10 +58,39 @@ apps.Application = Application;
+        appVersions.set('latest', application);
+      }
+    } else {
+      appVersions.set(application.version.toString(), application);
```
Can version be a non-string value?
I thought that we agreed on using the semver, which meant using strings as version values.

belochub commented on lib/applications.js:
```diff
@@ -56,10 +58,39 @@ apps.Application = Application;
+
+      // save latest version to fill missing latest versions later
+      const latestApp = latestApps.get(application.name);
+      if (!latestApp || application.version > latestApp.version) {
```
Same here, it is incorrect to compare semver versions like that.

belochub commented on lib/connection.js:
```diff
@@ -112,15 +112,24 @@ class Connection extends EventEmitter {
+    let target;
+    if (typeof app === 'string') {
+      target = app;
+    } else if (!app.version) {
```
You should also add a possibility to provide `app` as a string containing both name and version like that: `appName@1.0.0`.

lundibundi commented on lib/applications.js:
```diff
@@ -9,11 +9,13 @@ module.exports = apps;
 // your needs.
 //   name - application name
 //   api - application API
+//   version - api version of 'name' application
```
Sure.

lundibundi commented:
@belochub as discussed I've added support for semver in application versions. Also I've updated tests to check that functionality. Btw I've changed that comment message (api parameter description) event though this PR doesn't directly modify API but it is related to the API, so I think it's okay to put those changes here.
@metarhia/jstp-core PTAL.

belochub commented on lib/server.js:
```diff
@@ -24,6 +26,15 @@ const initServer = function(
   }
 
+  // versions cached for efficient search when provided version is a range
+  this._cachedVersions = {};
```
In my opinion, it is better to use `Object.create(null)` or even `Map` here.

lundibundi replied:
Oh, I again forgot about map, thanks.

belochub commented on lib/server.js:
```diff
@@ -84,6 +95,28 @@ class Server {
+  _getApplication(name, version) {
+    const appVersions = this.applications.get(name);
+    if (!appVersions) return null;
+    if (version) {
```
You can invert the condition here to avoid a lot of indentation below.

aqrln commented on lib/connection.js:
```diff
@@ -209,16 +221,21 @@ class Connection extends EventEmitter {
-      )
-    };
+    const packet = {};
+    if (Array.isArray(target)) {
```
I don't like the idea of making this function polymorphic. It is a hot function, and passing arguments of different types will mess up with V8's ICs.

lundibundi replied:
@aqrln I guess I'll split this into 2 functions with one accepting an array and the other value, what do you think?

aqrln replied:
Yeah, that would work if we needed it at all. See my [other comment](https://github.com/metarhia/jstp/pull/231/files#r126623600).

aqrln replied:
Why not always expect an array here though?

belochub replied:
@aqrln, to avoid creation of an array with one element in some cases, it should probably be better for performance.

lundibundi replied:
Also I'd add that in most cases it is a single argument, as handshake is the only place with array and it's performed only once per connection.

lundibundi replied:
@aqrln also what about this one?

aqrln replied:
@lundibundi based on the conversation above, I though there was consensus about it. I have approved splitting this into two functions above, and neither me nor anyone else has objected to your last comment, so feel free to pick either approach.

aqrln commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,177 @@
+  }
+};
+
+const interfacesVLatest = {
```
Wait... do I get it right that `latest` is a version by itself and not just a pointer to the latest semver version (`2.0.0` in this case)? That sounds wrong to me. `latest` and the like are tags, they must not be independent versions.

lundibundi replied:
No, It was made to check specific behavior (when app has no version it's latest version). If every app has a version then the biggest version will be the latest, see https://github.com/metarhia/jstp/pull/231/files/66c24103eef081bdb56c42696a400182502634f5#diff-13dd866a30aedfb1fab7c3d9da9446ceR162.

belochub commented:
@aqrln, ping.

aqrln commented on lib/connection.js:
```diff
@@ -112,15 +113,26 @@ class Connection extends EventEmitter {
   // Send a handshake packet over the connection
-  //   appName - name of an application to connect to
+  //   app - string or object, application to connect to as 'name' or
+  //         'name@version' or { name, version }
```
Do we really need all the three of these? No strong opinion on it, but I'd rather make `name@version` the only format.

lundibundi replied:
Same here, I don't really care. @metarhia/jstp-core somebody wants to leave the third option here?

belochub replied:
Yeah, why not?

lundibundi replied:
No more comments means 1 for, 0 against, hence we left it as is.

aqrln commented on lib/connection.js:
```diff
@@ -112,15 +113,26 @@ class Connection extends EventEmitter {
+    if (typeof app === 'string') {
+      const [name, version] = app.split('@');
+      if (semver.valid(version)) target = [name, version];
+      else target = name;
```
I'd throw an error here instead of silently discarding an invalid version.

lundibundi replied:
@aqrln Yeah, that will be better.

aqrln replied:
@lundibundi this is still unaddressed.

lundibundi replied:
@aqrln as you can see in the comments below I'm waiting to get your approval to make all changes at once as I don't want to perform rebase multiple times for naught.

aqrln commented on lib/connection.js:
```diff
@@ -112,15 +113,26 @@ class Connection extends EventEmitter {
+    } else if (!app.version) {
+      target = app.name;
+    } else {
+      target = [app.name, app.version];
```
Can you please stick to either a string (as in `{ handshake: [0, 'app@latest'] }`) or an array (as in `{ handshake: [0, 'app', 'latest'] }`) in both cases? As the parsed messages are plain objects, the current approach may suffer from the constantly changing object shapes.

belochub replied:
@aqrln, as far as I understand, first option you provided isn't possible, because it is always an array, check out `_createPacket` function. Sometimes version can just be omitted in the array, meaning that latest version should be used, and packet will look like this:
`{ handshake: [0, 'app'] }`.

aqrln replied:
Ah, I see. Right.

aqrln commented on lib/server.js:
```diff
@@ -24,6 +26,16 @@ const initServer = function(
+  applications.forEach((appVersions, appName) => {
+    const versions = Array.from(appVersions)
+      .filter(app => app[0] !== 'latest')
+      .map(version => [new semver.SemVer(version[0]), version[1]])
```
Can you please use destructuring in the left-hand side of the lambda to give some descriptive names for these values (i.e., `([something, somethingElse]) => [new semver.SemVer(something), somethingElse]`)?

Aside from that, why do `app` and `version` identifiers refer to the same objects in this and previous lines?

belochub replied:
@aqrln, there's comment on line 111 in this file.

aqrln replied:
It doesn't help much. The file is read top to bottom, and these variable names are somewhat misleading. Both `app` and `version` at lines 33 and 34 are actually arrays of `[version, app]`, AFAIU.

belochub replied:
Yes, but here it is obvious, because of line 32: this is an array created from the map with versions as keys, and applications as values, there is no need to make a destruction assignment here.
However, I do agree that arguments in lambdas should be renamed.

lundibundi replied:
@belochub @aqrln Also I see no particular reason to specifically emphasize arguments with destructuring here and it should be sufficient to rename `.filter` argument to `version` to be consistent, bit I don't really care so it's up to you.

aqrln replied:
This is still unaddressed.

lundibundi replied:
@aqrln but there is no approval of your to my proposed changes?

aqrln replied:
@lundibundi I would still prefer `([version, app]) => ...`, but if you and @belochub are against it, then at least making the names consistent works for me too.

aqrln commented on lib/server.js:
```diff
@@ -84,6 +96,29 @@ class Server {
+        if (range.test(version[0])) return version[1];
+      }
+    } catch (error) {
+      // TODO (lundibundi): after loggers are implemented write using them
```
Should it return a new error code instead of `ERR_APP_NOT_FOUND` in this case? 🤔

lundibundi replied:
@aqrln technically it's still missing application :smiley:, but it may indeed be clearer to actually add another error.

lundibundi replied:
@aqrln there is no response here as well.

aqrln replied:
@lundibundi there haven't been any changes requested as well ;)

I'm fine with `ERR_APP_NOT_FOUND`.

belochub commented:
@lundibundi, can you address the comments and rebase this on master, please?

lundibundi commented:
@belochub I've already answered the comments and haven't seen any response since then, that's why I'm waiting.

belochub commented:
@lundibundi, you've agreed to rename the arguments (https://github.com/metarhia/jstp/pull/231#discussion_r126699371) and you haven't done it.
Also you haven't answered this https://github.com/metarhia/jstp/pull/231#discussion_r126623048 at all.

lundibundi commented:
@belochub  https://github.com/metarhia/jstp/pull/231#discussion_r126699371 - yes but I thought we should get the approval of @aqrln as he was the one who suggested to change it and not in the way we decided to. As for the https://github.com/metarhia/jstp/pull/231#discussion_r126623048 - you've answered  on that one, there is nothing to add.

belochub commented:
I haven't answered https://github.com/metarhia/jstp/pull/231#discussion_r126623048, you've probably confused it with some other comment.

lundibundi commented:
@belochub oh yeah, I confused it with the one below, thanks.

nechaido commented:
@aqrln ping.

nechaido commented:
@lundibundi ping see https://github.com/metarhia/jstp/pull/231#issuecomment-315419755.

lundibundi commented:
@nechaido As you can see in the comment right below the one you referenced I've already answered on the comments and waiting for the decision. As for the rebase, as I said before I want to finish it all at once (after addressing comments) and see no particular reason to rebase now.

nechaido commented:
@lundibundi as far as I can see neither https://github.com/metarhia/jstp/pull/231#discussion_r126623048 nor https://github.com/metarhia/jstp/pull/231#discussion_r126627605 a resolved. Correct me I am wrong.

belochub commented:
@nechaido, @lundibundi is waiting for @aqrln to answer the comments you mentioned.

nechaido commented:
@belochub it just seems to be that @lundibundi already agreed to make requested changes.

belochub commented:
@nechaido, he just wants to rebase and upload all of the changes at once as he said in https://github.com/metarhia/jstp/pull/231#issuecomment-316675765.

aqrln commented on lib/applications.js:
```diff
@@ -56,10 +64,39 @@ apps.Application = Application;
+  // set latest versions of apps without explicit latest version
+  latestApps.forEach((application, appName) => {
+    const appVersions = index.get(appName);
+    if (!appVersions.get('latest')) appVersions.set('latest', application);
```
Nit: I'd prefer it to be multi-line.

aqrln commented on package-lock.json:
```diff
@@ -713,7 +713,7 @@
     "color-support": {
```
Can you please revert the `package-lock.json` changes, rebase onto the current `master` and reinstall `semver` with the latest npm?

belochub commented on lib/applications.js:
```diff
@@ -54,10 +65,41 @@ class Application {
+  // set latest versions of apps without explicit latest version
+  latestApps.forEach((application, appName) => {
+    const appVersions = index.get(appName);
+    if (!appVersions.get('latest')) {
```
You should use `has` method here instead of `get` because you don't need the value, just checking if it exists.

belochub commented on lib/connection.js:
```diff
@@ -109,15 +110,38 @@ class Connection extends EventEmitter {
+    let name, version;
+    if (typeof app === 'string') {
+      [name, version] = app.split('@');
+      if (version && !semver.valid(version)) {
```
Shouldn't this method allow providing a range instead of exact version, if you enabled support for the ranges on the server side (`_getApplication` method in `lib/server.js`)?

belochub commented on lib/simple-connect-policy.js:
```diff
@@ -6,9 +6,10 @@
 //
 module.exports = class SimpleConnectPolicy {
-  constructor(login, password) {
+  constructor(login, password, version) {
```
Why did you choose to provide version here instead of passing it to the `connect` method?

aqrln replied:
Good catch.

lundibundi replied:
Hm, I thought in the simplest case user doesn't need to change the version dynamically and if they do, they will implement their own connect policy with such behavior. But yeah, I forgot about appName in `connect`, so I'll add version there.

belochub commented on lib/connection.js:
```diff
@@ -120,8 +121,14 @@ class Connection extends EventEmitter {
-      if (version && !semver.valid(version)) {
-        throw new Error('Invalid semver version');
+      if (version && !semver.validRange(version)) {
+        const error =  new Error('Invalid semver version range');
```
Redundant space here.

belochub commented on lib/connection.js:
```diff
@@ -120,8 +121,14 @@ class Connection extends EventEmitter {
       [name, version] = app.split('@');
-      if (version && !semver.valid(version)) {
-        throw new Error('Invalid semver version');
+      if (version && !semver.validRange(version)) {
```
Also, you should add a check for the version provided inside of an object, now you just assign its value even if it's incorrect (on line 135).

belochub commented on lib/simple-connect-policy.js:
```diff
@@ -6,21 +6,21 @@
   // You can get client object provided upon connection creation
   // with connection.client.
+  //   app - string or object, application to connect to as 'name' or
+  //         'name@version' or { name, version }
```
You should specify the constraints for `version` like you did in `lib/connection.js` on lines 114-115.

belochub commented on test/node/api-versions.js:
```diff
@@ -74,10 +73,7 @@ test.test('must handle specific versions', (test) => {
-      connectPolicy: new jstp.SimpleConnectPolicy(null, null, '1'),
-    };
-    jstp.net.connect(app.name, client, port, (error, conn) => {
+    jstp.net.connect(app.name + '@1', null, port, (error, conn) => {
```
You know, that's the case for which passing app argument as an object, that I wanted to stay in this PR in https://github.com/metarhia/jstp/pull/231#discussion_r126700322, would come in handy.
This also goes for all of the similar occurrences in the tests.

belochub commented on lib/connection.js:
```diff
@@ -364,7 +417,9 @@ class Connection extends EventEmitter {
     const applicationName = packet.handshake[1];
-    const application = this.server.applications[applicationName];
+    const applicationVersion = packet.handshake[2];
+    const application =
```
I am not sure about the style of this, didn't we change styling for such cases?

belochub commented on lib/net.js:
```diff
@@ -55,8 +55,8 @@ module.exports = {
-    connectAndInspect(appName, client, interfaces, ...options),
+  connect: (app, client, ...options) =>
+    connect(app, client, ...options),
+  connectAndInspect: (app, client, interfaces, ...options) =>
```
You should also rename this argument in `lib/cli/command-processor.js`.

aqrln commented on lib/applications.js:
```diff
@@ -1,15 +1,26 @@
-  constructor(name, api) {
-    this.name = name;
+  constructor(name, api, version) {
+    [this.name, this.version] = name.split('@');
```
Should it split from the right instead of from the left?

aqrln replied:
I mean, `@metarhia/application@1.0` is a valid package name, so I'd expect it to be parsed as an application name correctly too.

belochub replied:
@aqrln, in such cases you can provide an object with fields `name` and `version` set to corresponding values.

aqrln replied:
@belochub I don't find it a valid justification. Why provide two ways to do the same thing with one of them being subtly broken? Either the npm-ish format must work according to users' expectations, or the `{ name, version }` schema must be the only available contract.

lundibundi replied:
@aqrln and in the above example will the user have application name as `@metarhia/application`? As we don't support scopes (they mean nothing in our implementation) is there any reason to support that? But anyway it's easy to implement so it's okay with me.

belochub replied:
@aqrln, by the way, are you sure that `@metarhia/application@1.0` is a valid package name?
Trying to create the package with such name using `npm` leads to the error:
`Sorry, name can only contain URL-friendly characters.`

aqrln replied:
Did you try to make `@1.0` a part of the package name or...?

aqrln commented on lib/connection.js:
```diff
@@ -109,15 +110,46 @@ class Connection extends EventEmitter {
+  handshake(app, login, password, callback) {
+    let name, version;
+    if (typeof app === 'string') {
+      [name, version] = app.split('@');
```
Ditto.

nechaido commented on lib/applications.js:
```diff
@@ -1,15 +1,27 @@
 class Application {
-  constructor(name, api) {
-    this.name = name;
+  constructor(name, api, version) {
```
I'd prefer to swap parameters `api` and `version` but it's up to you.

lundibundi replied:
Even though I agree that it will look better first of all it will not be compatible and second in case version is specified in the `name` you will have to pass `null` to the version parameter, because api will come after it.

nechaido commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,206 @@
+
+    const port = server.address().port;
+
+    // latest
```
I'd prefer this to be a subtest.

nechaido commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,206 @@
+      });
+    });
+
+    // v1
```
Ditto.

nechaido commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,206 @@
+      });
+    });
+
+    // v2
```
Ditto.

nechaido commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,206 @@
+
+    const port = server.address().port;
+
+    // ^1.0.0
```
Ditto.

nechaido commented on test/node/api-versions.js:
```diff
@@ -0,0 +1,206 @@
+      });
+    });
+
+    // > v1.0.0
```
Ditto.

belochub commented:
Landed in https://github.com/metarhia/jstp/commit/9116125e6dce8ab5a99519790693f37a5660937e.
